libs/micronaut-newrelic/src/main/java/com/agorapulse/micronaut/newrelic/BeanIntrospectionEventPayloadExtractor.java

```diff
@@ -44,10 +45,23 @@ public <E> Map<String, Object> extractPayload(E event) {
         Map<String, Object> map = new HashMap<>(propertyNames.length - 1);
 
         for (String name : propertyNames) {
-            introspection.getProperty(name)
-                .flatMap(p -> Optional.ofNullable(p.get(event)))
-                .map(v -> (v instanceof Boolean || v instanceof Number) ? v : String.valueOf(v))
-                .ifPresent(v -> map.put(name, v));
+            Optional<BeanProperty<E, Object>> property = introspection.getProperty(name);
+            if (property.isEmpty()) {
+                continue;
+            }
+            boolean isFlatten = property.get().getAnnotationMetadata().isAnnotationPresent(Flatten.class.getName());
+            if (!isFlatten) {
+                map.put(name, getValueWithSupportedType(property.get().get(event)));
+                continue;
+            }
+            if (property.get().get(event) instanceof Map additionalData) {
+                if (additionalData.keySet().stream().anyMatch(key -> !(key instanceof String))) {
+                    throw new IllegalArgumentException("@Flatten annotated getter must return Map<String, Object> but found a non String key in " + additionalData);
```

Comment:
can't we also support POJOs recursively? It would allow more readable event object where you know what additional data are supported.

Reply:
has discussed, it should only be used to provide dynamic fields in the event